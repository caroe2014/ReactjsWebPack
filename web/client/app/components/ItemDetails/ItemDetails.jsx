// Copyright Â© 2018-2019 Agilysys NV, LLC.  All Rights Reserved.  Confidential Information of Agilysys NV, LLC.

import React, { Component } from 'react';
import styled from 'styled-components';
import ItemCount from 'web/client/app/components/AddItem';
import PropTypes from 'prop-types';
import { Flex, Box, Text, Button } from 'rebass';
import ItemModifier from 'web/client/app/components/ItemModifier/ItemModifier';
import DynamicImage from 'web/client/app/components/DynamicImage';
import NotificationBar from 'web/client/app/components/NotificationBar';
import { Trans } from 'react-i18next';
import i18n from 'web/client/i18n';
import { currencyLocaleFormat } from 'web/client/app/utils/NumberUtils';
import SpecialInstructionInput from 'web/client/app/components/SpecialInstructionInput';

const Container = styled(Flex)`
  align-items: center;
  flex-wrap: wrap;
  flex-direction: column;
  margin: 0;
  width: 100%;
`;

const ChildBox = styled(Box)`
  max-width: 720px;
  border-bottom: none;
  border-top: none;
  margin-top: 120px;
  padding-top: 20px;
  width: 100%;
`;

const ItemDetailsContainer = styled(Box)`
  border: solid 1px ${props => props.theme.colors.itemBorder};
  padding: 20px;
  background-color: ${props => props.theme.colors.light};
  margin-bottom: 0px;
  &:focus {
    outline: none;
    box-shadow: none;
  }
`;

const TextContainer = styled(Text)`
  font-size: ${props => props.theme.fontSize.nm};
  padding-top: 15px;
  color:${props => props.theme.colors.textGrey};
  &:focus {
    outline: none;
    box-shadow: none;
  }
`;

const PrimaryTextContainer = styled(Text)`
  font-size: ${props => props.theme.fontSize.nm};
  padding-top: 15px;
  color:${props => props.theme.colors.primaryTextColor};
  &:focus {
    outline: none;
    box-shadow: none;
  }
`;

const DescriptionContainer = styled(Text)`
  font-size: ${props => props.theme.fontSize.nm};
  padding-top: 15px;
  color: ${props => props.theme.colors.secondaryTextColor};
  &:focus {
    outline: none;
    box-shadow: none;
  }
`;

const PriceText = styled(Text)`
  font-size: ${props => props.theme.fontSize.lg};
  padding-top: 12px;
  padding-left: 20px;
  font-weight:bold;
  color:${props => props.theme.colors.primaryTextColor};
`;

const TileContent = styled(Text)`
 color: ${props => props.theme.colors.secondaryTextColor};
 margin: 10px 10px
   ${props => props.theme.spacing.none} ${props => props.theme.spacing.none};
 font-size: 14px
 border: 0.5px solid lightgrey;
 border-radius: 12px;
 padding: 0px 5px;
 height:24px;
 white-space:nowrap;
 display: flex;
 align-items: center;
 ${props => props.theme.mediaBreakpoints.mobile`display: none;`};
 &:focus {
    outline: none;
    box-shadow: none;
  }
`;
TileContent.displayName = 'TileContent';

const CartContent = styled(Flex)`
  flex-direction: row;
  margin: ${props => props.theme.spacing.md} ${props => props.theme.spacing.none}
    ${props => props.theme.spacing.none} ${props => props.theme.spacing.none};
  justify-content: space-between;
`;

const QuantityContainer = styled(Flex)`
  visibility :hidden;
  opacity:0;
`;

const ImageContainer = styled(Box)`
  width: 100%;
  height: 15em;
  min-height: 15em;
  overflow-y: hidden;
`;

const TileContainer = styled(Flex)`
  flex-direction: row;
  &:focus {
    outline: none;
    box-shadow: none;
  }
`;

const TagParent = styled(Flex)`
  flex-direction: row;
  flex-wrap: wrap;
`;

const IconTileBox = styled(Flex)`
  display: inline-block;
  vertical-align: middle;
  justify-contet:center;
  align-items: center;
  margin-right: 10px;
  margin-top: 10px;
  white-space: nowrap;
  color: ${props => props.theme.colors.secondaryTextColor};
  ${props => props.theme.mediaBreakpoints.desktop`display: none;`};
  ${props => props.theme.mediaBreakpoints.tablet`display: none;`};
  ${props => props.theme.mediaBreakpoints.mobile`display: block;`};
`;

const IconTileText = styled(Text)`
  color: ${props => props.theme.colors.itemTile};
  font-size: 14px;
  margin-top: 2px;
  &:focus {
    outline: none;
    box-shadow: none;
  }
`;

const fontStyle = {
  fontSize: '1em',
  display: 'inline-block',
  verticalAlign: 'middle',
  marginRight: '8px'
};

const CartAmount = styled(Text)`
  flex-direction: row;
  color:${props => props.theme.colors.primaryTextColor};
  font-size: 18px;
  font-weight:bold;
  &:focus {
    outline: none;
    box-shadow: none;
  }
`;

const ModifierContainer = styled(Flex)`
  flex-direction: column;
  margin-left: -20px;
  margin-right: -20px;
  margin-top: 8px;
`;

const SpecialInstructionContainer = styled(Flex)`
  flex-direction: column;
  margin-left: -20px;
  margin-right: -20px;
  margin-top: 8px;
`;

const ModifierHeader = styled(Flex)`
  width: 100%;
  min-height: 40px;
  overflow-y: hidden;
  background-color:${props => props.theme.colors.disableBackground};
  font-weight:bold;
  align-items: center;
  padding-left:20px;
  color:${props => props.theme.colors.primaryTextColor};
  &:focus {
    outline: none;
    box-shadow: none;
  }
`;

const ButtonContainer = styled(Box)`
  margin:20px 0px 0px 0px;
`;

const AddcartButton = styled(Button)`
  background: ${props => props.theme.colors.buttonControlColor};
  color: ${props => props.theme.colors.buttonTextColor};
  cursor: pointer;
  font-weight: 500;
  font-size: 16px;
  float: right;
  width: 100%;
  height: 50px;
  padding: ${props => props.theme.spacing.sm} ${props => props.theme.spacing.nm};
  &:hover {
    box-shadow: 0px 0px 10px 2px lightgrey;
    -webkit-box-shadow: 0px 0px 10px 2px lightgrey;
    -moz-box-shadow: 0px 0px 10px 2px lightgrey;
  }
  ${props => {
    if (props.disablestate) {
      return `
      opacity: 0.5;
      visibility: visible;
      cursor: default;
      `;
    }
  }}
  text-align: center;
  margin: 0 auto;
  border: none;
  border-radius: 0;
  &:focus {
    outline: none;
    box-shadow: none;
  }
`;

const ResetButton = styled(Button)`
  background: ${props => props.theme.colors.light};
  border: 0.5px solid ${props => props.theme.colors.buttonControlColor};
  color: ${props => props.theme.colors.buttonControlColor};
  cursor: pointer;
  font-weight: 500;
  font-size: 16px;
  float: right;
  width: 100%;
  height: 50px;
  padding: ${props => props.theme.spacing.sm} ${props => props.theme.spacing.nm};
  &:hover {
    box-shadow: 0px 0px 10px 2px lightgrey;
    -webkit-box-shadow: 0px 0px 10px 2px lightgrey;
    -moz-box-shadow: 0px 0px 10px 2px lightgrey;
  }
  text-align: center;
  margin: 0 auto;
  border-radius: 0;
  &:focus {
    outline: none;
    box-shadow: none;
  }
`;

class ItemDetails extends Component {

  constructor (props) {
    super(props);
    let initAmount = this.props.selectItem ? this.props.selectItem.amount : 0;
    this.state = {
      totalAmount: initAmount,
      displayTotal: initAmount,
      addonTotal: 0,
      totalQuantity: 1,
      resetKey: 0,
      selectedModifiers: [],
      showValidationState: false,
      modifiersAreValid: false,
      invalidModifiers: [],
      openNotification: false
    };
    this.onQuantityChange = this.onQuantityChange.bind(this);
    this.onResetDefault = this.onResetDefault.bind(this);
    this.onAddToCart = this.onAddToCart.bind(this);
    this.onAddModifier = this.onAddModifier.bind(this);
    this.setValidModifiersState = this.setValidModifiersState.bind(this);
    this.notify = this.notify.bind(this);
    this.specialInstructionInput = this.specialInstructionInput.bind(this);
  }
  componentWillUpdate (newProps) {
    if (newProps.selectItem && JSON.stringify(this.props.selectItem) !== JSON.stringify(newProps.selectItem)) {
      this.setState({ totalAmount: newProps.selectItem.amount, displayTotal: newProps.selectItem.amount });
    }
  }
  onQuantityChange (quantity) {
    const { selectItem } = this.props;
    let totalAmount = (quantity * selectItem.amount).toFixed(2);
    let displayTotal = this.calculateDisplayTotal(totalAmount, quantity, this.state.addonTotal);
    this.setState({ totalQuantity: quantity, totalAmount: totalAmount, displayTotal: displayTotal });
  }
  onResetDefault () {
    const { selectItem } = this.props;
    this.setState({
      totalQuantity: 1,
      totalAmount: selectItem.amount,
      displayTotal: selectItem.amount,
      addonTotal: 0,
      resetKey: this.state.resetKey + 1,
      selectedModifiers: []
    });
  }
  onAddModifier (modifierData) {
    let { totalAmount } = this.state;
    let { addOnAmount } = modifierData.modifiersList;
    let displayTotal = this.calculateDisplayTotal(totalAmount, this.state.totalQuantity, addOnAmount);
    this.setState({
      displayTotal: displayTotal,
      addonTotal: addOnAmount,
      selectedModifiers: modifierData.selectedModifiers
    });
  }
  onAddToCart () {
    this.setState({ showValidationState: true });
    if (this.state.modifiersAreValid || !this.props.selectItem.modifiers) {
      let item = { ...this.props.selectItem };
      item.count = this.state.totalQuantity;
      if (this.state.selectedModifiers && this.state.selectedModifiers.length > 0) {
        item.selectedModifiers = Object.assign([], this.state.selectedModifiers);
      }
      if (this.state.splInstruction !== undefined) {
        item.splInstruction = this.state.splInstruction;
      }
      this.props.handleClick && this.props.handleClick(item);
    } else {
      let invalidModifierNames = '';
      const appendModifierName = function (m, i, a) {
        invalidModifierNames += ` ${m.description}`;
        if (i === a.length - 2) {
          invalidModifierNames += i18n.t('ITEM_DETAILS_MISSING_REQUIRED_AND');
        } else if (i !== a.length - 1) {
          invalidModifierNames += ',';
        }
      };
      this.state.invalidModifiers.forEach(appendModifierName);
      this.notify(i18n.t('ITEM_DETAILS_MISSING_REQUIRED', { invalidModifierNames: invalidModifierNames, interpolation: { escapeValue: false } }));// eslint-disable-line max-len
      invalidModifierNames && document.getElementById('tab-list') && document.getElementById('tab-list').focus();
    }
  }
  setValidModifiersState (modifiersAreValid, invalidModifiers) {
    this.setState({ modifiersAreValid, invalidModifiers });
  }
  calculateDisplayTotal (totalAmount, quantity, addonTotal) {
    return (parseFloat(totalAmount) + parseFloat(quantity * addonTotal)).toFixed(2);
  }

  notify (notificationText) {
    this.setState({ notificationText });
  }

  specialInstructionInput (splInstructionstext) {
    this.setState({ splInstruction: splInstructionstext });
  }

  canShowAttribute (itemDisplayList, itemProperty) {
    return itemDisplayList.length === 0 ? true : (itemDisplayList.find(itemDisplayParam => itemDisplayParam === itemProperty)); // eslint-disable-line max-len
  }

  render () {
    const { selectItem, currencyDetails, disableAddToCart, specialInstructions, itemDisplayList } = this.props;
    const { notificationText } = this.state;

    const showDescription = this.canShowAttribute(itemDisplayList, 'description');
    const showLabels = this.canShowAttribute(itemDisplayList, 'labels');
    const showImage = this.canShowAttribute(itemDisplayList, 'image');
    const showPrice = this.canShowAttribute(itemDisplayList, 'price');

    return (
      <Container>
        {selectItem &&
          <ChildBox className='box-container' width={[1, 0.65]} >
            {selectItem.image && showImage &&
              <ImageContainer className='image-container'>
                <DynamicImage
                  className='image'
                  bgSize='cover'
                  style={{ width: '100%' }}
                  src={selectItem.image}
                />
              </ImageContainer>
            }
            <ItemDetailsContainer className='details-container'>
              {showDescription && selectItem.description &&
                <DescriptionContainer
                  className='description'
                  pt={[15]}
                  pb={[10]}
                  tabIndex={0}
                  aria-label={selectItem.description}
                >
                  {selectItem.description}
                </DescriptionContainer>
              }
              <TileContainer className='description-tab' flexWrap='wrap'>
                <TagParent className='tileTab' width={[1, 0.6, 0.6, 0.6]} alignItems='center'>
                  {showLabels && selectItem.tagNames && selectItem.tagNames.map((item) => (
                    <React.Fragment key={item}>
                      <TileContent className='icon-tile' role='img' tabIndex={0} aria-label={item}>
                        <i
                          className={`agilysys-icon-${item.toLowerCase().replace(/\s/g, '-')}`}
                          style={fontStyle}
                          alt=''
                        />
                        {item}
                      </TileContent>
                      <IconTileBox className='icon-tile'
                        role={item.id !== 'calories' && 'img'}
                        tabIndex={0} aria-label={item.id !== 'calories' ? item : item.value}>
                        {item.id === 'calories'
                          ? <IconTileText>{item.value}</IconTileText>
                          : <i
                            className={`agilysys-icon-${item.toLowerCase().replace(/\s/g, '-')}`}
                            style={fontStyle}
                            alt=''
                          />
                        }
                      </IconTileBox>
                    </React.Fragment>
                  ))}
                </TagParent>
                {showPrice &&
                  <TileContainer
                    width={[1, 0.4, 0.4, 0.4]}
                    alignItems='center'
                    justifyContent={['space-between', 'flex-end']}
                    tabIndex={0}
                    aria-label={`${i18n.t('ITEM_LABEL')}${i18n.t('ITEM_DETAILS_BASE_PRICE')} 
                    ${currencyLocaleFormat(selectItem.amount, currencyDetails)}`}
                  >
                    <PrimaryTextContainer
                      className='title base-price-label'
                      fontWeight='bold'
                      children={<Trans i18nKey='ITEM_DETAILS_BASE_PRICE' />}
                    />
                    <PriceText className='base-price-text'>
                      {currencyLocaleFormat(selectItem.amount, currencyDetails)}
                    </PriceText>
                  </TileContainer>
                }
              </TileContainer>
              {selectItem.modifiers && <ModifierContainer>
                <ModifierHeader className='header-text'
                  tabIndex={0}
                  aria-label={i18n.t('ITEM_DETAILS_CUSTOMIZE')}
                  children={<Trans i18nKey='ITEM_DETAILS_CUSTOMIZE' />} />
                <ItemModifier
                  key={this.state.resetKey}
                  modifiers={selectItem.modifiers}
                  handlerFromParent={this.onAddModifier}
                  showValidationState={this.state.showValidationState}
                  onValidate={this.setValidModifiersState}
                  onNotify={this.notify}
                  currencyDetails={currencyDetails}
                />
                <NotificationBar
                  text={notificationText}
                  actionButtonText={<Trans i18nKey='ITEM_DETAILS_GOT_IT' />}
                  onActionClick={() => { this.setState({ notificationText: undefined }); }}
                  open={notificationText !== undefined}
                />
              </ModifierContainer>
              }

              {specialInstructions && specialInstructions.featureEnabled &&
                <SpecialInstructionContainer>
                  <SpecialInstructionInput
                    key={this.state.resetKey}
                    headerText={specialInstructions.headerText}
                    instructionText={specialInstructions.instructionText}
                    characterLimit={specialInstructions.characterLimit}
                    handleSplInstruction={this.specialInstructionInput}
                  />
                </SpecialInstructionContainer>
              }
              <CartContent alignItems='center'>
                <QuantityContainer alignItems='center' >
                  <TextContainer
                    p={['0px 15px 0px 0px !important']}
                    children={<Trans i18nKey='ITEM_DETAILS_QUANTITY' />}
                  />
                  <ItemCount
                    key={this.state.resetKey}
                    min={1}
                    max={99}
                    width={100}
                    onInputChange={this.onQuantityChange}
                  />
                </QuantityContainer>
                <CartAmount className='total-amount-text' tabIndex={0}
                  aria-label={`${currencyLocaleFormat(this.state.displayTotal, currencyDetails)}`}>
                  {currencyLocaleFormat(this.state.displayTotal, currencyDetails)}
                </CartAmount>
              </CartContent>
              <TileContainer flexWrap='wrap'>
                <ButtonContainer width={[1, 1 / 2, 1 / 2, 1 / 2]} pr={[0, 2]}>
                  <ResetButton
                    className='reset-button button'
                    type='submit'
                    onClick={this.onResetDefault}
                    children={<Trans i18nKey='ITEM_DETAILS_RESET' />}
                    tabIndex={0}
                  />
                </ButtonContainer>
                <ButtonContainer width={[1, 1 / 2, 1 / 2, 1 / 2]} pl={[0, 2]}>
                  <AddcartButton
                    className='add-to-cart-button'
                    disablestate={disableAddToCart ? 1 : 0}
                    type='submit'
                    onClick={!disableAddToCart ? this.onAddToCart : undefined}
                    tabIndex={0}
                    children={<Trans i18nKey='ITEM_ADD_TO_CART' />}
                  />
                </ButtonContainer>
              </TileContainer>
            </ItemDetailsContainer>
          </ChildBox>
        }
      </Container>
    );
  }
}

ItemDetails.propTypes = {
  selectItem: PropTypes.shape({
    image: PropTypes.string.isRequired,
    name: PropTypes.string.isRequired,
    description: PropTypes.string.isRequired,
    tagNames: PropTypes.arrayOf(PropTypes.string.isRequired).isRequired
  }),
  currencyDetails: PropTypes.object.isRequired,
  itemDisplayList: PropTypes.array
};

export default ItemDetails;
