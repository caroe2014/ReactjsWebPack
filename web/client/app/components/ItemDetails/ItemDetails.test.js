// Copyright © 2018-2019 Agilysys NV, LLC.  All Rights Reserved.  Confidential Information of Agilysys NV, LLC.

import React from 'react';
import {
  shallow,
  mount,
  configure
} from 'enzyme';
import toJson from 'enzyme-to-json';
import Adapter from 'enzyme-adapter-react-16';
import {
  ThemeProvider
} from 'styled-components';
import theme from 'web/client/theme';
import i18n from 'i18next';
import ItemDetails from './ItemDetails';
import jsdom from 'jsdom';

configure({
  adapter: new Adapter()
});

jest.mock('web/client/i18n', () => ({
  t: (key) => {
    return key;
  }
}));
jest.mock('react-i18next', () => ({
  // this mock makes sure any components using the translate HoC receive the t function as a prop
  withTranslation: () => Component => {
    Component.defaultProps = { ...Component.defaultProps, t: (key) => key };
    return Component;
  }
}));

jest.mock('react-i18next', () => ({
  Trans: 'div'
}));

/* global describe, it, expect, beforeEach */
/* eslint-disable max-len */
global.window = new jsdom.JSDOM().window;
global.document = window.document;

describe('ItemDetails', () => {
  let mockProps = {
    image: 'https://www.ndtv.com/cooks/images/grilled.sandwich.jpg',
    displayText: 'Mc Alister’s Club',
    amount: '12.00',
    description: 'Organic and Natural sandwiches with delicious sauce made to required sizes and shapes as per wish.',
    tagNames: ['Veg', '160 cal'],
    modifiers: {
      'modifiers': [
        {
          'id': 'bread',
          'description': 'BREAD',
          'type': 'radio',
          'options': [
            {
              'id': 'wheatbaguette',
              'description': 'Wheat Baguette',
              'selected': false
            },
            {
              'id': 'wheatwrap',
              'description': 'Wheat wrap',
              'selected': false
            },
            {
              'id': 'spinachwrap',
              'description': 'Spinach wrap',
              'selected': false
            }
          ]
        },
        {
          'id': 'addon',
          'description': 'ADD ON',
          'type': 'checkbox',
          'options': [
            {
              'id': 'Fountain Drink',
              'description': 'Fountain Drink',
              'selected': true,
              'amount': '2.50',
              'options': [
                {
                  'id': 'Large',
                  'description': 'Large',
                  'selected': true,
                  'amount': '2.50'
                },
                {
                  'id': 'Small',
                  'description': 'Small',
                  'selected': false,
                  'amount': '1.50'
                }
              ]
            },
            {
              'id': 'Small Soup',
              'description': 'Small Soup',
              'selected': false,
              'amount': '1.50'
            },
            {
              'id': 'Large Soup',
              'description': 'Aged Cheddar',
              'selected': true,
              'amount': '1.50'
            },
            {
              'id': 'Swiss',
              'description': 'Swiss',
              'selected': true,
              'amount': '1.50'
            }
          ]
        }
      ],
      'addOnAmount': 0
    }
  };
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<ItemDetails itemDisplayList={['labels', 'description', 'image']} selectItem={mockProps}/>);
  });
  it('does a snapshot test', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should render without throwing an error', () => {
    expect(wrapper.exists()).toBe(true);
  });

  it('should render without throwing an error on mount', () => {
    const mountWrapper = mount(<ThemeProvider theme={theme}><ItemDetails itemDisplayList={['labels', 'description', 'image']} selectItem={mockProps} disableAddToCart/></ThemeProvider>);
    expect(mountWrapper.exists()).toBe(true);
  });

  it('checks whether image prop is recevied', () => {
    let imageComponent = wrapper.find('DynamicImage');
    expect(imageComponent.props().src).toBe(mockProps.image);
  });

  it('checks whether description prop is recevied', () => {
    let description = wrapper.find('.description');
    expect(description.props().children).toBe(mockProps.description);
  });

  it('checks whether tags prop is recevied', () => {
    let tags = wrapper.find('TileContent');
    expect(tags.length).toBe(mockProps.tagNames.length);
  });

  it('calls add to cart when clicked whether tags prop is recevied', () => {
    let mockFn = jest.fn();
    let OnAddWrapper = shallow(<ItemDetails itemDisplayList={['labels', 'description', 'image']} selectItem={mockProps} handleClick={mockFn}/>);
    const spy = jest.spyOn(OnAddWrapper.instance(), 'onAddToCart');
    OnAddWrapper.instance().forceUpdate();
    OnAddWrapper.setState({ modifiersAreValid: true });
    OnAddWrapper.find('.add-to-cart-button').simulate('click');
    expect(spy).toHaveBeenCalled();
    spy.mockClear();
  });

  it('calls add to cart when clicked whether tags prop is recevied', () => {
    let OnAddWrapper = shallow(<ItemDetails itemDisplayList={['labels', 'description', 'image']} selectItem={mockProps}/>);
    const spy = jest.spyOn(OnAddWrapper.instance(), 'onAddToCart');
    OnAddWrapper.instance().forceUpdate();
    OnAddWrapper.setState({ modifiersAreValid: false });
    OnAddWrapper.find('.add-to-cart-button').simulate('click');
    expect(spy).toHaveBeenCalled();
    spy.mockClear();
  });

  it('checks whether reset button work properly', () => {
    let itemCount = wrapper.find('AddItem').dive();
    let plusButton = itemCount.find('Styled(IconButton)').last();
    plusButton.props().onClick();
    itemCount.update();
    wrapper.update();
    let resetButton = wrapper.find('.button');
    resetButton.props().onClick();
    wrapper.update();
    let resetAmount = wrapper.find('Styled(Text)').last().props().children;
    expect(`$${mockProps.amount}`).toBe(resetAmount);
  });

  describe('when Image prop is not given', () => {
    beforeEach(() => {
      mockProps.image = '';
      wrapper = shallow(<ItemDetails itemDisplayList={['labels', 'description', 'image']} selectItem={mockProps}/>);
    });
    it('checks whether image prop is not recevied', () => {
      let imageComponent = wrapper.find('DynamicImage');
      expect(imageComponent.length).toBe(0);
    });
  });

  describe('when image is not present in itemDisplayList prop', () => {
    beforeEach(() => {
      wrapper = shallow(<ItemDetails itemDisplayList={['labels', 'description']} selectItem={mockProps}/>);
    });
    it('checks whether image is rendered', () => {
      let imageComponent = wrapper.find('DynamicImage');
      expect(imageComponent.length).toBe(0);
    });
  });

  describe('when modifiers data given', () => {
    it('checks whether modifiers shown', () => {
      let modifier = wrapper.find('ItemModifier');
      expect(modifier.length).toBe(1);
    });
    it('checks total amount with selected modifiers', () => {
      let totalAddOnAmount = Number(mockProps.amount);
      mockProps.modifiers.modifiers.map((tabs) => {
        const selectedOptions = tabs.options.filter(item => (item.selected === true && item.amount));
        selectedOptions.map((option) => {
          totalAddOnAmount = totalAddOnAmount + Number(option.amount);
        });
      });
      let itemdetails = wrapper;
      itemdetails.find('ItemModifier').dive().find('.option').first().childAt(0).simulate('click');
      itemdetails.update();
      expect(itemdetails.find('Styled(Text)').last().props().children).toBe(`$${totalAddOnAmount.toFixed(2)}`);
    });
  });

  describe('componentWillUpdate', () => {
    let newMockProps = {
      image: 'https://www.ndtv.com/cooks/images/grilled.sandwich.jpg',
      displayText: 'sMc Alister’s Club',
      amount: '12.00',
      description: 'Organic and Natural sandwiches with delicious sauce made to required sizes and shapes as per wish.',
      tagNames: ['Veg', '160 cal'],
      modifiers: {
        'modifiers': [
          {
            'id': 'bread',
            'description': 'BREAD',
            'type': 'radio',
            'options': [
              {
                'id': 'wheatbaguette',
                'description': 'Wheat Baguette',
                'selected': false
              },
              {
                'id': 'wheatwrap',
                'description': 'Wheat wrap',
                'selected': false
              },
              {
                'id': 'spinachwrap',
                'description': 'Spinach wrap',
                'selected': false
              }
            ]
          },
          {
            'id': 'addon',
            'description': 'ADD ON',
            'type': 'checkbox',
            'options': [
              {
                'id': 'Fountain Drink',
                'description': 'Fountain Drink',
                'selected': true,
                'amount': '2.50',
                'options': [
                  {
                    'id': 'Large',
                    'description': 'Large',
                    'selected': true,
                    'amount': '2.50'
                  },
                  {
                    'id': 'Small',
                    'description': 'Small',
                    'selected': false,
                    'amount': '1.50'
                  }
                ]
              },
              {
                'id': 'Small Soup',
                'description': 'Small Soup',
                'selected': false,
                'amount': '1.50'
              },
              {
                'id': 'Large Soup',
                'description': 'Aged Cheddar',
                'selected': true,
                'amount': '1.50'
              },
              {
                'id': 'Swiss',
                'description': 'Swiss',
                'selected': true,
                'amount': '1.50'
              }
            ]
          }
        ],
        'addOnAmount': 0
      }
    };
    it('updates with newprops', () => {
      wrapper.setProps({ selectItem: newMockProps });
    });
  });
});
