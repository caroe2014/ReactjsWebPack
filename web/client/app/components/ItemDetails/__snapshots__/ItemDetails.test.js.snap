// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ItemDetails does a snapshot test 1`] = `
<Styled(Flex)>
  <Styled(Box)
    className="box-container"
    width={
      Array [
        1,
        0.65,
      ]
    }
  >
    <Styled(Box)
      className="image-container"
    >
      <DynamicImage
        bgSize="cover"
        className="image"
        src="https://www.ndtv.com/cooks/images/grilled.sandwich.jpg"
        style={
          Object {
            "width": "100%",
          }
        }
      />
    </Styled(Box)>
    <Styled(Box)
      className="details-container"
    >
      <Styled(Text)
        aria-label="Organic and Natural sandwiches with delicious sauce made to required sizes and shapes as per wish."
        className="description"
        pb={
          Array [
            10,
          ]
        }
        pt={
          Array [
            15,
          ]
        }
        tabIndex={0}
      >
        Organic and Natural sandwiches with delicious sauce made to required sizes and shapes as per wish.
      </Styled(Text)>
      <Styled(Flex)
        className="description-tab"
        flexWrap="wrap"
      >
        <Styled(Flex)
          alignItems="center"
          className="tileTab"
          width={
            Array [
              1,
              0.6,
              0.6,
              0.6,
            ]
          }
        >
          <TileContent
            aria-label="Veg"
            className="icon-tile"
            role="img"
            tabIndex={0}
          >
            <i
              alt=""
              className="agilysys-icon-veg"
              style={
                Object {
                  "display": "inline-block",
                  "fontSize": "1em",
                  "marginRight": "8px",
                  "verticalAlign": "middle",
                }
              }
            />
            Veg
          </TileContent>
          <Styled(Flex)
            aria-label="Veg"
            className="icon-tile"
            role="img"
            tabIndex={0}
          >
            <i
              alt=""
              className="agilysys-icon-veg"
              style={
                Object {
                  "display": "inline-block",
                  "fontSize": "1em",
                  "marginRight": "8px",
                  "verticalAlign": "middle",
                }
              }
            />
          </Styled(Flex)>
          <TileContent
            aria-label="160 cal"
            className="icon-tile"
            role="img"
            tabIndex={0}
          >
            <i
              alt=""
              className="agilysys-icon-160-cal"
              style={
                Object {
                  "display": "inline-block",
                  "fontSize": "1em",
                  "marginRight": "8px",
                  "verticalAlign": "middle",
                }
              }
            />
            160 cal
          </TileContent>
          <Styled(Flex)
            aria-label="160 cal"
            className="icon-tile"
            role="img"
            tabIndex={0}
          >
            <i
              alt=""
              className="agilysys-icon-160-cal"
              style={
                Object {
                  "display": "inline-block",
                  "fontSize": "1em",
                  "marginRight": "8px",
                  "verticalAlign": "middle",
                }
              }
            />
          </Styled(Flex)>
        </Styled(Flex)>
      </Styled(Flex)>
      <Styled(Flex)>
        <Styled(Flex)
          aria-label="ITEM_DETAILS_CUSTOMIZE"
          className="header-text"
          tabIndex={0}
        >
          <div
            i18nKey="ITEM_DETAILS_CUSTOMIZE"
          />
        </Styled(Flex)>
        <ItemModifier
          handlerFromParent={[Function]}
          key="0"
          modifiers={
            Object {
              "addOnAmount": 0,
              "modifiers": Array [
                Object {
                  "description": "BREAD",
                  "id": "bread",
                  "options": Array [
                    Object {
                      "description": "Wheat Baguette",
                      "id": "wheatbaguette",
                      "selected": false,
                    },
                    Object {
                      "description": "Wheat wrap",
                      "id": "wheatwrap",
                      "selected": false,
                    },
                    Object {
                      "description": "Spinach wrap",
                      "id": "spinachwrap",
                      "selected": false,
                    },
                  ],
                  "type": "radio",
                },
                Object {
                  "description": "ADD ON",
                  "id": "addon",
                  "options": Array [
                    Object {
                      "amount": "2.50",
                      "description": "Fountain Drink",
                      "id": "Fountain Drink",
                      "options": Array [
                        Object {
                          "amount": "2.50",
                          "description": "Large",
                          "id": "Large",
                          "selected": true,
                        },
                        Object {
                          "amount": "1.50",
                          "description": "Small",
                          "id": "Small",
                          "selected": false,
                        },
                      ],
                      "selected": true,
                    },
                    Object {
                      "amount": "1.50",
                      "description": "Small Soup",
                      "id": "Small Soup",
                      "selected": false,
                    },
                    Object {
                      "amount": "1.50",
                      "description": "Aged Cheddar",
                      "id": "Large Soup",
                      "selected": true,
                    },
                    Object {
                      "amount": "1.50",
                      "description": "Swiss",
                      "id": "Swiss",
                      "selected": true,
                    },
                  ],
                  "type": "checkbox",
                },
              ],
            }
          }
          onNotify={[Function]}
          onValidate={[Function]}
          showValidationState={false}
        />
        <NotificationBar
          actionButtonText={
            <div
              i18nKey="ITEM_DETAILS_GOT_IT"
            />
          }
          onActionClick={[Function]}
          open={false}
        />
      </Styled(Flex)>
      <Styled(Flex)
        alignItems="center"
      >
        <Styled(Flex)
          alignItems="center"
        >
          <Styled(Text)
            p={
              Array [
                "0px 15px 0px 0px !important",
              ]
            }
          >
            <div
              i18nKey="ITEM_DETAILS_QUANTITY"
            />
          </Styled(Text)>
          <AddItem
            key="0"
            max={99}
            min={1}
            onInputChange={[Function]}
            width={100}
          />
        </Styled(Flex)>
        <Styled(Text)
          aria-label="$12.00"
          className="total-amount-text"
          tabIndex={0}
        >
          $12.00
        </Styled(Text)>
      </Styled(Flex)>
      <Styled(Flex)
        flexWrap="wrap"
      >
        <Styled(Box)
          pr={
            Array [
              0,
              2,
            ]
          }
          width={
            Array [
              1,
              0.5,
              0.5,
              0.5,
            ]
          }
        >
          <Styled(Button)
            className="reset-button button"
            onClick={[Function]}
            tabIndex={0}
            type="submit"
          >
            <div
              i18nKey="ITEM_DETAILS_RESET"
            />
          </Styled(Button)>
        </Styled(Box)>
        <Styled(Box)
          pl={
            Array [
              0,
              2,
            ]
          }
          width={
            Array [
              1,
              0.5,
              0.5,
              0.5,
            ]
          }
        >
          <Styled(Button)
            className="add-to-cart-button"
            disablestate={0}
            onClick={[Function]}
            tabIndex={0}
            type="submit"
          >
            <div
              i18nKey="ITEM_ADD_TO_CART"
            />
          </Styled(Button)>
        </Styled(Box)>
      </Styled(Flex)>
    </Styled(Box)>
  </Styled(Box)>
</Styled(Flex)>
`;
